mm/mm_heap/mm_sem.c

```diff
@@ -102,7 +103,7 @@ bool mm_takesemaphore(FAR struct mm_heap_s *heap)
       return false;
     }
 #if defined(CONFIG_BUILD_FLAT) || defined(__KERNEL__)
-  else if (sched_idletask())
+  else if (sched_idletask() || up_interrupt_context())
```

Comment:
It looks like this path is also taken when `CONFIG_DEBUG_MM` is disabled and we are in an interrupt context.  

Maybe if we pull the interrupt handling aside the code would look simpler, like:

```c
#if (defined(CONFIG_BUILD_FLAT) || defined(__KERNEL__)))
 
  /* Check current environment */

  if (up_interrupt_context())
    {
#  ifdef CONFIG_DEBUG_MM
      return _SEM_TRYWAIT(&heap->mm_semaphore) >= 0;
#  else
      /* Can't take semaphore in the interrupt handler */
      return false;
#endif
    }
  else
#endif

```